Replace hard limits to highscore saving with configurable limits

BWagener opened the pull request:
Currently only the top 16 scores on a song will be saved. If a player achieves below 16 already saved scores their score will not be saved.

### What does this PR do?

* replaces the constants used as conditions to decide if a score should be saved with configurable values
  * specifically the limit of scores saved by song and the minimum required score for saving (unchanged from previous value of 2000 points)
* store all scores above 2000 by default
* fixes the maxLines logic that determines how many lines should be output to the list of scores

### Motivation

This comes in handy when you are playing with players who score lower on average and currently can never save their scores, once a song has 16 scores saved for it.

twollgam commented on game/hiscore.cc:
```diff
@@ -26,7 +24,7 @@ bool Hiscore::reachedHiscore(unsigned score, SongId songid, unsigned short level
 		if (elem.level != level) continue;
 		if (score > elem.score) return true;
-		if (++position >= MaximumStoredScores) return false;
+		if (config["game/highscore_limit_stored_scores"].b() && ++position >= config["game/highscore_maximum_stored_scores"].ui()) return false;
```
Some time ago we decided to write the if-/else-blocks into a new line. Don't know if we updated the wiki. But if you change a line, put the blocks into an own line please.

twollgam commented on game/hiscore.cc:
```diff
@@ -43,6 +41,12 @@ void Hiscore::addHiscore(HiscoreItem&& item) {
 	m_hiscore.insert(std::move(item));
 }
 
+void Hiscore::loadHiscore(HiscoreItem&& item) {
```
`load` seams not the right term for this function. Why not `add` or if you want to differentiate from the other add-functions `store` or `put`?

twollgam commented on game/screen_songs.cc:
```diff
@@ -356,7 +356,8 @@ void ScreenSongs::draw() {
 	auto const datetimeFormat = config["game/datetime_format"].so();
-	auto const maxLines = 8;
+	// allows for example for 16 entries for a single track
+	auto const maxLines = 17;
```
Did you try it and does all lines are displayed correct on the screen? (Use window mode and different windows sizes)

Cause layout rendering is not that good in performous.

twollgam commented on game/screen_songs.cc:
```diff
@@ -379,17 +380,23 @@ std::string ScreenSongs::getHighScoreText() const {
 	auto stream = std::stringstream();
 	auto n = 0;
 	for (auto const& [track, scores]: scoresByTrack) {
+		if (n + 2 >= maxLines) break; // don't print a track if no scores for it would be printed
```
New line for the `break` please.

twollgam commented on game/screen_songs.cc:
```diff
@@ -379,17 +380,23 @@ std::string ScreenSongs::getHighScoreText() const {
 	for (auto const& [track, scores]: scoresByTrack) {
+		if (n + 2 >= maxLines) break; // don't print a track if no scores for it would be printed
 		stream << track << ":\n";
+		n++;
```
Use the prefix notation where possible. Should be the default. The postfix notation could lead to performance problems in some cases.

twollgam commented on game/screen_songs.cc:
```diff
@@ -379,17 +380,23 @@ std::string ScreenSongs::getHighScoreText() const {
 			timeFormatter(stream, score.unixtime);
 			stream << "\n";
+			n++;
+			if (n >= maxLines) break;
```
New line.

twollgam commented on game/screen_songs.cc:
```diff
@@ -379,17 +380,23 @@ std::string ScreenSongs::getHighScoreText() const {
+			if (n >= maxLines) break;
 		}
+
+		if (n >= maxLines) break;
```
New line.

twollgam commented on game/screen_songs.cc:
```diff
@@ -379,17 +380,23 @@ std::string ScreenSongs::getHighScoreText() const {
-			break;
-		}
+		n++;
+		if (n >= maxLines) break;
```
New line.

twollgam reviewed (changes requested):
All in all it's good, but some minor changes are needed.

And there is one question belonging layout things.

BWagener commented:
> All in all it's good, but some minor changes are needed.
> 
> And there is one question belonging layout things.

Great, I updated the formatting and prefix notation related stuff. Let me know what you think about the replies regarding the layout and function name.

twollgam commented:
> > All in all it's good, but some minor changes are needed.
> > And there is one question belonging layout things.
> 
> Great, I updated the formatting and prefix notation related stuff. Let me know what you think about the replies regarding the layout and function name.

Yes I would, but I see only the code changes. Did you save the comments?

BWagener commented:
Do these links work?

https://github.com/performous/performous/pull/983#discussion_r1573269556
https://github.com/performous/performous/pull/983#discussion_r1573274340

twollgam commented:
> Do these links work?
> 
> [#983 (comment)](https://github.com/performous/performous/pull/983#discussion_r1573269556) [#983 (comment)](https://github.com/performous/performous/pull/983#discussion_r1573274340)

It's strange. In the popups of the links I see the start of your comment but a click doesn't lead me to them. And I cannot find them anywhere in the pr. 

@Baklap4 could this be a rights problem?

BWagener commented:
That's odd, I'm on mobile but I'll try to copy the comments content here:

#1 regarding the function name
I figured it would fit since this is only used when loading the songs initially from db, put and store sound more generic.

Which one do you prefer?

#2 regarding the length of the displayed high scores
It doesn't seem to change with screen size.

This is 640x480:
image
It's the same for example on 1080p and 1440p.

The last two lines could overlap with a long song title. The same is true for 14 song entries. With 12 it only very slightly reaches into the song title box:
image

Personally I don't mind the list reaching that far down but I understand if you prefer to keep it smaller. We could also make the list length configurable.

Or we could align the scores to the vertical top instead of center, then there would be quite a bit more space.

BWagener commented:
My comments regarding some thoughts I had on specific sections of the code ended up being their own review, does that show up for you?

It shows up for me, almost at the top of this thread. And my answers to your comments are there as well, for some reason. I'm pretty sure I used the reply function to respond to your review comments for those, though. Quite strange.

twollgam commented:
> #1 regarding the function name I figured it would fit since this is only used when loading the songs initially from db, put and store sound more generic.
> 
> Which one do you prefer?

Ok I understand your point. But a function name should always describe what the function is doing, not where it should be used ;-) You'll never know who could use the function where in the future.

So I prefer `addSong` like the others.

BWagener commented:
> But maybe I miss a thing. If you (@BWagener) are able to relate to call it simple `addHiscore`.

Yeah I think it'd clash with the existing `addHiscore` function with the same signature. What do you think about `addHiscoreUnconditionally`?

twollgam commented:
> > But maybe I miss a thing. If you (@BWagener) are able to relate to call it simple `addHiscore`.
> 
> Yeah I think it'd clash with the existing `addHiscore` function with the same signature. What do you think about `addHiscoreUnconditionally`?

Yes, I agree completely 👍 Good suggestion. Take it and I will finish the review. Can you write some test hints to the pr description? Then we search someone how will test it or I will test it on Windows 11 next week.

BWagener commented:
> > > But maybe I miss a thing. If you (@BWagener) are able to relate to call it simple `addHiscore`.
> > 
> > 
> > Yeah I think it'd clash with the existing `addHiscore` function with the same signature. What do you think about `addHiscoreUnconditionally`?
> 
> Yes, I agree completely 👍 Good suggestion. Take it and I will finish the review. Can you write some test hints to the pr description? Then we search someone how will test it or I will test it on Windows 11 next week.

Great, I updated the PR description and I updated the function name in the latest commit. I updated the function docs to reflect what the functions currently do - the addSong docs seemed a bit outdated so I updated that too.

twollgam reviewed (dismissed):
LGTM

BWagener commented:
I'd also in favor of changing the default minimum score required for saving.

I can imagine removing it completely might not be desirable though, I haven't tested how the score saving screen behaves in that case.
 
How about 100?

ooshlablu commented:
@BWagener Can you try rebasing this on current master to try to get the Window build to pass?

BWagener commented:
> @BWagener Can you try rebasing this on current master to try to get the Window build to pass?

@ooshlablu this seems to be up-to-date with the latest master
